tests/virt_autotest/kubevirt_tests_agent.pm

```diff
@@ -0,0 +1,91 @@
+# SUSE's openQA tests
+#
+# Copyright 2022 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Run kubevirt testing
+# Maintainer: Nan Zhang <nan.zhang@suse.com>
+
+use base multi_machine_job_base;
+use strict;
+use warnings;
+use testapi;
+use lockapi;
+use utils;
+
+sub run {
+    my ($self) = @_;
+    my $ip_out = script_output('ip route show | grep -Eo "src\s+([0-9.]*)\s+" | head -1 | cut -d\' \' -f 2', 30);
+
+    set_var('AGENT_IP', $ip_out);
+    bmwqemu::save_vars();
+
+    # Mark ready state
+    mutex_create('DST_READY_TO_START');
+
+    my $server_ip = $self->get_var_from_child("SERVER_IP");
+
+    rke2_agent_setup($self, $server_ip);
+}    
+# Mutex lock/unlock functionality
+sub set_mutex_lock {
+    my ($self, $lock_signal) = @_;
+
+    mutex_lock($lock_signal);
+    mutex_unlock($lock_signal);
+}
+
+sub rke2_agent_setup {
+    my ($self, $server_ip) = @_;
+
+    systemctl('stop apparmor.service');
+    # systemctl('stop SuSEfirewall2.service');
+    assert_script_run('ssh-keygen -b 2048 -t rsa -q -N "" -f ~/.ssh/id_rsa') unless (! -e "~/.ssh/id_rsa");
+    exec_and_insert_password("ssh-copy-id -o StrictHostKeyChecking=no -i /root/.ssh/id_rsa.pub root\@$server_ip");
+
+    # Ensure SUSE certificates are installed on the node
+    ensure_ca_certificates_suse_installed();
+
+    # Install downstream kubevirt packages
+    zypper_call('in -n kubernetes1.18-client') if (script_run('rpmquery kubernetes1.18-client'));
+
+    # RKE2 deployment on agent node
+    my $rke_ver = get_var('INSTALL_RKE2_VERSION');
+    if ($rke_ver) {
+        assert_script_run("curl -sfL https://get.rke2.io | INSTALL_RKE2_VERSION=$rke_ver INSTALL_RKE2_TYPE=agent sh -");
+    } else {
+        assert_script_run('curl -sfL https://get.rke2.io | INSTALL_RKE2_TYPE=agent sh -');
+    }
+	systemctl('enable rke2-agent.service');
+
+    # Wait for rke2-server service to be ready
+    $self->set_mutex_lock('RKE2_SERVER_START_READY');
+    my $server_node_token = script_output("ssh root\@$server_ip cat /var/lib/rancher/rke2/server/node-token");
+
+    # Configure rke2-agent service
+    assert_script_run('mkdir -p /etc/rancher/rke2/');
+    assert_script_run("echo 'server: https://$server_ip:9345' > /etc/rancher/rke2/config.yaml");
+    assert_script_run("echo 'token: $server_node_token' >> /etc/rancher/rke2/config.yaml");
+
+    # Start rke2-agent service ready
+    systemctl('start rke2-agent.service');
+    mutex_create('RKE2_AGENT_START_READY');
+
+    assert_script_run("mkdir -p ~/.kube; scp root\@$server_ip:/etc/rancher/rke2/rke2.yaml ~/.kube/config");
+    assert_script_run("sed -ie 's/127.0.0.1/$server_ip/' ~/.kube/config");
+    assert_script_run('kubectl get nodes');
+
+    # Wait for creating registries to be ready
+    $self->set_mutex_lock('CREATE_REGISTRIES_READY');
+
+    # Restart RKE2 service and check the service is active well after restart
+    systemctl('restart rke2-agent.service');
+    check_service_status();
+}
```

Comment:
I think communication between agent and server can be improved a little bit:
1. Do not give the other end a signal to proceed if something goes wrong on itself.
2. Only give the other end a signal to proceed if everything goes well on itself.
3. Consider signal alarm to the other end if something goes wrong on itself.